File: agent/api/copy_render.py
"""HTTP surface for the On-Demand Copy Renderer (Round 2).

Product → Benefit → Duration → Generate 5 suggestions → lock/regenerate to target
→ finalize → prepare N READY packages. Thin handlers delegate to
``copy_render_service``.

Auth (amendment 8): EVERY endpoint requires an authenticated human session (401);
mutations additionally require ``products.update`` (403). Benefit copy is strictly
HYBRID/FACELESS (enforced in the service). prepare-selected returns READY packages
only — it NEVER enqueues production, runs video, or touches the Copy-V2 binding.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query

from agent.models.copy_render_v1 import (
    CreateCopyRenderSessionRequest,
    GenerateSuggestionsRequest,
    SetVisualConfigRequest,
    UpdateTargetRequest,
)
from agent.security.access_control import get_current_auth_context
from agent.services import copy_render_service as svc

router = APIRouter(prefix="/copy-render", tags=["copy-render"])

_MUTATION_PERMISSION = "products.update"


def _raise(exc: svc.CopyRenderError):
    raise HTTPException(
        status_code=exc.status_code,
        detail={"error": exc.code, "message": exc.message, "details": exc.details,
                "provider_calls": exc.provider_calls},
    ) from exc


def _require_actor():
    actor = get_current_auth_context()
    if actor is None:
        raise HTTPException(
            status_code=401,
            detail={"error": "AUTHENTICATION_REQUIRED", "message": "An authenticated session is required."},
        )
    return actor


def _require_mutation_actor():
    actor = _require_actor()
    if _MUTATION_PERMISSION not in actor.permission_codes:
        raise HTTPException(
            status_code=403,
            detail={"error": "PERMISSION_DENIED",
                    "message": f"This action requires the {_MUTATION_PERMISSION} permission."},
        )
    return actor


# --------------------------------------------------------------------------
# session lifecycle
# --------------------------------------------------------------------------
@router.post("/sessions")
async def create_session(req: CreateCopyRenderSessionRequest) -> dict[str, Any]:
    actor = _require_mutation_actor()
    try:
        return await svc.create_session(
            product_id=req.product_id, benefit_id=req.benefit_id, lane=req.lane,
            target_count=req.target_count, duration_seconds=req.duration_seconds,
            target_language=req.target_language, formula_id=req.formula_id,
            created_by=actor.user_id, avatar_id=req.avatar_id,
        )
    except svc.CopyRenderError as exc:
        _raise(exc)


@router.patch("/sessions/{session_id}/visual-config")
async def set_visual_config(session_id: str, req: SetVisualConfigRequest) -> dict[str, Any]:
    """Bind the governed presenter identity (Avatar Registry avatar) for HYBRID.

    Visual config only: provider-free, never triggers a copy provider call and
    never mutates generated copy text or lineage. FACELESS is avatar-exempt.
    """
    _require_mutation_actor()
    try:
        return await svc.set_visual_config(session_id, avatar_id=req.avatar_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


@router.get("/sessions/{session_id}")
async def get_session(session_id: str) -> dict[str, Any]:
    _require_actor()
    session = await svc.get_session(session_id)
    if session is None:
        raise HTTPException(
            status_code=404,
            detail={"error": "COPY_RENDER_SESSION_NOT_FOUND", "message": "Unknown session_id."},
        )
    return session


@router.patch("/sessions/{session_id}/target")
async def update_target(session_id: str, req: UpdateTargetRequest) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.update_target(session_id, req.target_count)
    except svc.CopyRenderError as exc:
        _raise(exc)


# --------------------------------------------------------------------------
# suggestions (the ONLY provider-spending endpoint; idempotent + single-flight)
# --------------------------------------------------------------------------
@router.post("/sessions/{session_id}/suggestions")
async def generate_suggestions(session_id: str, req: GenerateSuggestionsRequest) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.generate_suggestions(session_id, req.request_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


# --------------------------------------------------------------------------
# lock / unlock / finalize (ZERO provider calls)
# --------------------------------------------------------------------------
@router.post("/candidates/{candidate_id}/lock")
async def lock_candidate(candidate_id: str) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.lock_candidate(candidate_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


@router.post("/candidates/{candidate_id}/unlock")
async def unlock_candidate(candidate_id: str) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.unlock_candidate(candidate_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


@router.post("/sessions/{session_id}/finalize")
async def finalize_session(session_id: str) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.finalize_session(session_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


# --------------------------------------------------------------------------
# selected reads + prepare-selected (N READY packages; NO production/queue/video)
# --------------------------------------------------------------------------
@router.get("/sessions/{session_id}/selected")
async def selected(session_id: str) -> dict[str, Any]:
    _require_actor()
    try:
        return await svc.selected(session_id)
    except svc.CopyRenderError as exc:
        _raise(exc)


@router.post("/sessions/{session_id}/prepare-selected")
async def prepare_selected(session_id: str) -> dict[str, Any]:
    _require_mutation_actor()
    try:
        return await svc.prepare_selected(session_id)
    except svc.CopyRenderError as exc:
        _raise(exc)

File: dashboard/src/api/copyRender.ts
// On-Demand Copy Renderer (Round 2) API client.
// Imports ONLY ./client so this feature stays disjoint from the Copy Register V2
// module. SYSTEM OWNS STRUCTURE; AI ONLY STITCHES. A benefit-render candidate is a
// distinct authority — NEVER a V2 binding.
import { getAPI, patchAPI, postAPI } from './client'

export type CopyRenderLane = 'HYBRID' | 'FACELESS'
export type CopyRenderStatus = 'OPEN' | 'TARGET_COMPLETE' | 'FINALIZED' | 'STALE' | 'CANCELLED'
export type CandidateStatus = 'SHOWN' | 'LOCKED' | 'SKIPPED' | 'FINALIZED'

export interface CopyRenderStage {
  stage_key: string
  text: string
}

export interface CopyRenderCandidate {
  candidate_id: string
  status: CandidateStatus
  recipe_fingerprint: string
  text_digest: string
  batch_id: string
  artifact_id: string
  full_copy_text: string | null
  word_count: number | null
  stages: CopyRenderStage[]
}

export interface CopyRenderBatchView {
  batch_id: string
  batch_number: number
  request_id: string
  action: 'GENERATE' | 'REGENERATE'
  status: 'RESERVED' | 'RUNNING' | 'SHOWN' | 'FAILED'
  provider_calls: number | null
  cache_hit_count: number | null
  failure_code: string | null
}

export interface CopyRenderSession {
  session_id: string
  product_id: string
  benefit_id: string
  lane: CopyRenderLane
  duration_seconds: number
  target_language: string
  formula_id: string
  word_budget: number
  target_count: number
  locked_count: number
  status: CopyRenderStatus
  regenerate_enabled: boolean
  total_unique_capacity: number
  used_recipe_count: number
  remaining_unique_capacity: number
  candidates: CopyRenderCandidate[]
  batches: CopyRenderBatchView[]
  finalized_at: string | null
  // present only on the Generate/Regenerate response
  provider_calls?: number
  batch_id?: string
}

export interface PreparedPackage {
  candidate_id: string
  package_id?: string
  artifact_id?: string
  reused?: boolean
  status: 'READY' | 'PACKAGE_ERROR'
  execution_allowed?: boolean
  blockers?: string[]
  prompt_fingerprint?: string
  error?: string
  detail?: string
}

export interface PrepareResult {
  session_id: string
  lane: CopyRenderLane
  package_count: number
  packages: PreparedPackage[]
  enqueued: boolean
}

export interface SelectedResult {
  session_id: string
  status: CopyRenderStatus
  count: number
  selected: {
    candidate_id: string
    status: CandidateStatus
    recipe_fingerprint: string
    full_copy_text: string | null
    word_count: number | null
    stages: CopyRenderStage[]
  }[]
}

const BASE = '/api/copy-render'
const q = (v: string) => encodeURIComponent(v)

/** A fresh per-Generate idempotency key. The SAME logical request never burns two
 * provider calls; a retry after failure must use a NEW request_id. */
export function newRequestId(): string {
  const g = globalThis as { crypto?: { randomUUID?: () => string } }
  if (g.crypto?.randomUUID) return `req-${g.crypto.randomUUID()}`
  return `req-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 12)}`
}

export interface CreateSessionInput {
  product_id: string
  benefit_id: string
  lane: CopyRenderLane
  target_count: number
  duration_seconds: number
  target_language?: string
  formula_id?: string | null
  /** Governed Avatar Registry presenter for a presenter-led (HYBRID) session, so the
   * session carries the visual config the backend requires before prepare-selected
   * (COPY_RENDER_HYBRID_AVATAR_REQUIRED). Reuse the operator's already-selected
   * presenter — never a default/substitute. Omit for avatar-exempt lanes (FACELESS). */
  avatar_id?: string | null
}

export async function createSession(input: CreateSessionInput): Promise<CopyRenderSession> {
  return postAPI(`${BASE}/sessions`, input)
}

export async function getSession(sessionId: string): Promise<CopyRenderSession> {
  return getAPI(`${BASE}/sessions/${q(sessionId)}`)
}

export async function updateTarget(sessionId: string, targetCount: number): Promise<CopyRenderSession> {
  return patchAPI(`${BASE}/sessions/${q(sessionId)}/target`, { target_count: targetCount })
}

/** Generate or Regenerate 5 suggestions. Pass a fresh requestId (see newRequestId). */
export async function generateSuggestions(sessionId: string, requestId: string): Promise<CopyRenderSession> {
  return postAPI(`${BASE}/sessions/${q(sessionId)}/suggestions`, { request_id: requestId })
}

export async function lockCandidate(candidateId: string): Promise<CopyRenderSession> {
  return postAPI(`${BASE}/candidates/${q(candidateId)}/lock`, {})
}

export async function unlockCandidate(candidateId: string): Promise<CopyRenderSession> {
  return postAPI(`${BASE}/candidates/${q(candidateId)}/unlock`, {})
}

export async function finalizeSession(sessionId: string): Promise<CopyRenderSession> {
  return postAPI(`${BASE}/sessions/${q(sessionId)}/finalize`, {})
}

export async function getSelected(sessionId: string): Promise<SelectedResult> {
  return getAPI(`${BASE}/sessions/${q(sessionId)}/selected`)
}

/** Materialize N READY prompt packages from the FINALIZED selection. This NEVER
 * enqueues production, runs video, or touches the Copy Register V2 binding. */
export async function prepareSelected(sessionId: string): Promise<PrepareResult> {
  return postAPI(`${BASE}/sessions/${q(sessionId)}/prepare-selected`, {})
}
